orttraining/orttraining/training_ops/cpu/rnn/lstm_io_utils.cc

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "orttraining/training_ops/cpu/rnn/lstm_io_utils.h"
+
+namespace onnxruntime::lstm {
+
+namespace {
+
+template <typename T>
+rnn::detail::GemmWeights<T> LoadWeights(const Tensor* weights, const int index) {
+  // index represents the direction of the weight to be loaded.
+  // For example,
+  //   in a uni-directional lstm, index can only ever be 0.
+  //   in a bi-directional lstm, index 0 represents forward weights and index 1 represents backward weights
+  const auto& weights_shape = weights->Shape();
+  const auto* weights_data = weights->Data<T>();
+  const size_t weights_size_per_direction = SafeInt<size_t>(weights_shape[1]) * weights_shape[2];
+  return rnn::detail::GemmWeights<T>(index, weights_data, weights_size_per_direction, rnn::detail::PackedWeights());
+}
+
+void ValidateLSTMInputs(const Tensor* X, const Tensor* W, const Tensor* R, const Tensor* B,
+                        const Tensor* SL, const Tensor* H0, const Tensor* C0, const Tensor* P,
+                        const int directions, const int hidden_size) {
+  const auto& X_shape = X->Shape();
+  ORT_ENFORCE(X_shape.NumDimensions() == 3U, "Input X must have 3 dimensions only. Actual:", X_shape);
+  const int batch_size = narrow<int>(X_shape[1]);
+  const int input_size = narrow<int>(X_shape[2]);
+
+  const auto& W_shape = W->Shape();
+  ORT_ENFORCE(W_shape.NumDimensions() == 3U &&
+                  narrow<int>(W_shape[0]) == directions &&
+                  narrow<int>(W_shape[1]) == 4 * hidden_size &&
+                  narrow<int>(W_shape[2]) == input_size,
+              "Input W must have shape {", directions, ", ", 4 * hidden_size, ", ", input_size, "}. Actual:", W_shape);
+
+  const auto& R_shape = R->Shape();
+  ORT_ENFORCE(R_shape.NumDimensions() == 3U &&
+                  narrow<int>(R_shape[0]) == directions &&
+                  narrow<int>(R_shape[1]) == 4 * hidden_size &&
+                  narrow<int>(R_shape[2]) == hidden_size,
+              "Input R must have shape {", directions, ", ", 4 * hidden_size, ", ", hidden_size, "}. Actual:", R_shape);
+
+  if (B != nullptr) {
+    const auto& B_shape = B->Shape();
+    ORT_ENFORCE(B_shape.NumDimensions() == 2U &&
+                    narrow<int>(B_shape[0]) == directions &&
+                    narrow<int>(B_shape[1]) == 8 * hidden_size,
+                "Input B must have shape {", directions, ", ", 8 * hidden_size, "}. Actual:", B_shape);
+  }
+
+  ORT_ENFORCE(!SL, "Varying sequence lengths is not supported yet for LSTMTraining and LSTMGrad.");
```

Comment:
You're right. By varyling sequence lengths, I mean the input `sequence_lengths` as a tensor where each input tensor has a different sequence length.
I can see why this comment is not clear. I'll update the error message so that it is clearer.